kotlinx-coroutines-core/common/src/flow/operators/Share.kt

```diff
@@ -412,6 +412,29 @@ private class SubscribedSharedFlow<T>(
         sharedFlow.collect(SubscribedFlowCollector(collector, action))
 }
 
+/**
+ * Returns a flow that invokes the given [action] **after** this state flow starts to be collected
+ * (after the subscription is registered).
+ *
+ * The [action] is called before any value is emitted from the upstream
+ * flow to this subscription but after the subscription is established. It is guaranteed that all emissions to
+ * the upstream flow that happen inside or immediately after this `onSubscription` action will be
+ * collected by this subscription.
+ *
+ * The receiver of the [action] is [FlowCollector], so `onSubscription` can emit additional elements.
+ */
+public fun <T> StateFlow<T>.onSubscription(action: suspend FlowCollector<T>.() -> Unit): StateFlow<T> =
```

Comment:
Please also add `@ExperimentalCoroutinesApi`. We can be fairly certain that this API is benign, but still, better safe than sorry, and also, this gives us the formal permission to publish this in a minor release if needed.

Reply:
JIC: if `@ExperimentalCoroutinesApi` will be added, then current users who use `SharedFlow.onSubscription` on a variables of type `StateFlow` will start to receive an opt-in warning without a nice ability to use previous overload.

Reply:
They could do `.let { it as SharedFlow<T> }.onSubscription`. You raise a good point, though!